Issue #4732 url link tool

CruzeBlade opened the pull request:
Hi, I added a new tool to embed links into the document as requested in issue #4732. 
This is just an early draft of that feature with basic functionality. 

I'm sharing the current status to get feedback and see if there is a chance to merge this in the future, as this will cause compatibility issues with older versions of Xournal++.

Here is a short demo:
![ezgif com-video-to-gif](https://user-images.githubusercontent.com/37212628/228538271-f2eaf21a-e182-447d-b7f6-3e1e249decec.gif)

What still needs to be done:
- [x] Integrate with PDF export
- [x] Open URL by clicking in Browser

CruzeBlade committed:
Started implementing Link model class

CruzeBlade committed:
Display Link element on page

CruzeBlade committed:
Fixed issue with random isInEditing

CruzeBlade committed:
Corrected comments

CruzeBlade committed:
Added new item to toolbar for creating & editing links

CruzeBlade committed:
Update cursor on selecting link toolbar item

CruzeBlade committed:
Removed debugging std::couts

CruzeBlade committed:
Add Link to page when link toolitem selected

CruzeBlade committed:
Save and load links in and from file

CruzeBlade committed:
Removed debugging cout

CruzeBlade committed:
Replaced creating link element by couts

CruzeBlade committed:
Added icons for url tool

CruzeBlade committed:
Improved rendering of Links

CruzeBlade committed:
Started outsource link editing login into separate LinkEditor

CruzeBlade committed:
Removed LinkEditor references from Control class

CruzeBlade committed:
Draw links with underline

CruzeBlade committed:
Implemented LinkDialog for creating / editing link elements

CruzeBlade committed:
Added empty lines at file end

CruzeBlade committed:
Changed cursor icon for link tool

CruzeBlade committed:
Implemented proper widget resizing after text edit

CruzeBlade committed:
Added a red bounding box to link while editing

CruzeBlade committed:
Fixed rendering issue when resizing to smaller

CruzeBlade committed:
Implemented hover effects for link elements

CruzeBlade committed:
Open links in default application by ctrl + click

CruzeBlade committed:
Started improving link dialog

CruzeBlade committed:
LinkDialog resize with text input

CruzeBlade committed:
Tried to fix scaling

CruzeBlade committed:
Improved LinkDialog UI

CruzeBlade committed:
LinkDialog implemented layout toggle buttons

CruzeBlade committed:
Impl changing text alignment for link elements

CruzeBlade committed:
Fixed issue with LinkDialog text alignment buttons

CruzeBlade committed:
Renamed link attribute inEditing to highlighted

CruzeBlade committed:
LinkDialog change link font

CruzeBlade committed:
Changed link color

CruzeBlade committed:
Add selected prefix to url

CruzeBlade committed:
Show link url in tooltip while hovering

CruzeBlade committed:
Tried to fix mac build

CruzeBlade committed:
Fixed popover bug with mulitple links

CruzeBlade committed:
Store text alignment for link elements in xopp

CruzeBlade committed:
Fixed scaling issue

CruzeBlade committed:
Fixed issue: closing the dialog deleted  link

CruzeBlade commented:
After having work on this feature for the last couple of weeks, here is a small status update on what has changed since.

- Open the embedded link (normal click opens the editor dialog, and a ctrl + click will open the link in the default program) 
- Improved the link dialog editor
   - Automatically resizes the text field until a certain height
   - Changed text font and alignment options
   - link prefix combo box 
- "Hover" effect for the link element

Here are some images showing the current status:
| Improved dialog  | Hover effects |
|---|---|
| ![demo_improved_link_dialog](https://github.com/xournalpp/xournalpp/assets/37212628/e1674fca-8043-4539-a6f2-7211f1ec8a7a)  | ![demo_link_hover_effect](https://github.com/xournalpp/xournalpp/assets/37212628/ab31db96-0b3f-4ddb-b5dc-791c10146da5) |

Any feedback is greatly appreciated 😎!

netlimpopo commented:
>   * Open the embedded link (normal click opens the editor dialog, and a ctrl + click will open the link in the default program)
> 
>     * Improved the link dialog editor
Hello! 
Thanks for all the effort! This looks amazing hope it can get mefged. It would be a valueable contribution to he project!
I am no maintainer so I have bo say in it. 
In the past it has been noted in the forums though, that people with "touch devices only" need to be included as well (convertibles), so the programm gets more barrier free. They won't have a Ctrl-Key available when the transform their decive into a touch only one. 
Hence they won't be able to click the link, only edit the link. There have been similiar discussions on the implementation of the PDF Tools.  
#4673 back then "alt" was used, and later neglected. 
Regards N.

CruzeBlade commented:
Hi, thank you for the feedback. That is a very good point I haven't thought about. I will have to find a solution for that.

rolandlo reviewed (commented):
The tool looks great! I like it. Backwards compatibility is certainly something we need to consider carefully, but as it currently stands, files generated with the new version containing links can still be read by older versions, just the link elements are gone (and don't come back when saving). 

Another thing that has to be considered carefully is how the links can be activated and how they can be edited. I haven't thought it through carefully, but I tend to think that clicking on the link should activate it, not edit it. That would be consistent with PDF links with the PDF tool. Usually a link will need to be activated more often (and maybe by less versed people) than edited. Those who only use links may only have a touch device and stylus. Those who create links will usually have a keyboard, too, since otherwise how would they type the text. So CTRL+click (or similar) for editing (already created) links looks acceptable.

I hope this tool will eventually also allow to add internal links. It doesn't have to be done in this PR though.

I have briefly skimmed through the code and it looks nicely structured and readable. So thanks for putting all this effort into the new tool! It can bring Xournal++ a good step ahead.

CruzeBlade commented:
## Regarding how to open a link:

I'm personally not in favor of opening a link by simply clicking on it. Especially on touch devices, it is quite common to accidentally click somewhere (e.g. while scrolling). This would not only be annoying, but could potentially also be a security issue when somebody has embedded a malicious link. Therefore, I think it is very important that the user can see the full URL before opening it (hence the popover in the current version) and that there needs to be a mechanism to verify the user intentionally wants to open the link (hence the CTRL+click). But that's just my personal opinion on that topic.

My initial thought on how to improve the controls was:
* Double click for editing & creating
 -> In the link dialog, I would add a button to open the link
* Single click will highlight the link (see hover effects) 
 -> I would restyle the URL label as a link
 -> Clicking on the URL in the popover will open the link (as with pdf links)
* Single click + CTRL opens the link (as is)

That was just some first idea, and I'm open for discussion and other suggestions. 

## Regarding internal linking:

I would prefer to do this in a second PR, just to not further inflate the complexity of this PR. 

## Regarding backwards compatibility:

Does Xournal++ have a warning if users open a newer file with an older xournal++ version? If not something like this might be useful. (But out of scope for this PR)

CruzeBlade committed:
Started implementing new control interface

CruzeBlade committed:
Continue working on new user controls

CruzeBlade committed:
Finished implementing new controls for link tool

CruzeBlade committed:
Improved link popover

CruzeBlade committed:
Improved link tool controls

CruzeBlade committed:
Further control improvements

CruzeBlade committed:
Reposition popover when scrolling

CruzeBlade committed:
Make links clickable in pdf export

CruzeBlade commented:
I have now implemented the new controls as mentioned in the previous post.

Here a small demo showing the enhanced popover (I added some control hints for the user):
![Demo_new_controls](https://github.com/xournalpp/xournalpp/assets/37212628/8dcad4a7-f498-44f6-ac7b-85bd3a559f77)

I currently don't own any touch device on which I can run Xournal++. Therefore, it would be great if somebody else could help me out and test if everything works as intended with touch devices.

CruzeBlade committed:
Dynamically set max height dialog

CruzeBlade committed:
Improved automatically resizing of link dialog

CruzeBlade committed:
Removed debugging console logs

CruzeBlade commented:
I have now finalized the implementation, and therefore I will make this a pull request.

rolandlo commented:
I have tried out the PR once more and I really like how you solved the way of choosing between editing and following the link. Could you please rebase on master? I want to look at the code in detail. Also please take into account  #4763 for making dialogs GTK4 compatible.

One minor thing I found that could possibly be improved: The default value for the text field for the URL is the same regardless of the URI scheme. It doesn't make sense though to write an email to `xournalpp.github.io` for example. So maybe a different default could be used for different schemes.

CruzeBlade committed:
Different default values per URI scheme

CruzeBlade committed:
Fixed issue related to rebasing

Mr-DblH commented:
Is there any news here yet? I find this almost indispensable for teaching purposes.

rolandlo commented:
Closing this PR since #6841 was merged.
